renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
+        VK_NULL_HANDLE,
+        VK_NULL_HANDLE,
+        (uint32_t)asGeomData.size(),
+        asGeomData.data(),
+        VK_NULL_HANDLE,
+    };
 
-    // Copy the data using host-commands but from mapped memory
-    byte *mappedSrcBuffer = NULL;
-    VkResult vkr = ObjDisp(d)->MapMemory(Unwrap(d), Unwrap(initial.mem.mem), initial.mem.offs, size,
-                                         0, (void **)&mappedSrcBuffer);
-    CHECK_VKR(m_pDriver, vkr);
+    rdcarray<uint32_t> counts;
+    counts.reserve(asGeomData.size());
+    for(VkAccelerationStructureBuildRangeInfoKHR numPrims : buildRangeInfos)
+      counts.push_back(numPrims.primitiveCount);
 
-    VkCopyMemoryToAccelerationStructureInfoKHR copyInfo = {
-        VK_STRUCTURE_TYPE_COPY_MEMORY_TO_ACCELERATION_STRUCTURE_INFO_KHR};
-    copyInfo.src.hostAddress = mappedSrcBuffer;
-    copyInfo.dst = unwrappedAs;
-    copyInfo.mode = VK_COPY_ACCELERATION_STRUCTURE_MODE_DESERIALIZE_KHR;
-    ObjDisp(d)->CopyMemoryToAccelerationStructureKHR(Unwrap(d), VK_NULL_HANDLE, &copyInfo);
+    ObjDisp(d)->GetAccelerationStructureBuildSizesKHR(
+        Unwrap(d), VK_ACCELERATION_STRUCTURE_BUILD_TYPE_DEVICE_KHR, &sizeInfo, counts.data(),
+        &sizeResult);
   }
-  else
+  UpdateScratch(sizeResult.buildScratchSize);
+
+  // Build the AS
+  VkCommandBuffer cmd = m_pDriver->GetInitStateCmd();
+  if(cmd == VK_NULL_HANDLE)
   {
-    const VkBufferDeviceAddressInfo addrInfo = {VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO, NULL,
-                                                Unwrap(initial.buf)};
-    const VkDeviceAddress uploadBufAddr = ObjDisp(d)->GetBufferDeviceAddressKHR(Unwrap(d), &addrInfo);
-
-    VkCopyMemoryToAccelerationStructureInfoKHR copyInfo = {
-        VK_STRUCTURE_TYPE_COPY_MEMORY_TO_ACCELERATION_STRUCTURE_INFO_KHR};
-    copyInfo.src.deviceAddress = uploadBufAddr;
-    copyInfo.dst = unwrappedAs;
-    copyInfo.mode = VK_COPY_ACCELERATION_STRUCTURE_MODE_DESERIALIZE_KHR;
-    ObjDisp(d)->CmdCopyMemoryToAccelerationStructureKHR(Unwrap(cmd), &copyInfo);
+    RDCERR("Couldn't acquire command buffer");
+    return;
   }
 
-  VkMarkerRegion::End(cmd);
+  const VkAccelerationStructureBuildGeometryInfoKHR asGeomInfo = {
+      VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_GEOMETRY_INFO_KHR,
+      NULL,
+      asInfo->type,
+      asInfo->flags,
+      VK_BUILD_ACCELERATION_STRUCTURE_MODE_BUILD_KHR,
+      VK_NULL_HANDLE,
+      Unwrap(wrappedAS),
+      (uint32_t)asGeomData.size(),
+      asGeomData.data(),
+      NULL,
+      scratchAddressUnion,
+  };
+
+  const VkAccelerationStructureBuildRangeInfoKHR *pBuildInfo = buildRangeInfos.data();
+  ObjDisp(d)->CmdBuildAccelerationStructuresKHR(Unwrap(cmd), 1, &asGeomInfo, &pBuildInfo);
 
-  if(Vulkan_Debug_SingleSubmitFlushing())
-  {
-    m_pDriver->CloseInitStateCmd();
-    m_pDriver->SubmitCmds();
-    m_pDriver->FlushQ();
-  }
+  // We serialise the AS builds so we can have just a single scratch buffer and reuse it
```

Comment:
Serialising the AS builds should not require a full GPU sync, you should be able to prevent overlap in builds with the same scratch space with a barrier.